src/features/ScreenLock/ScreenLockProvider.tsx

```diff
@@ -84,57 +87,54 @@ class ScreenLockProvider extends Component<
       this.props.password &&
       !(this.props.encryptedDbState && this.props.encryptedDbState.data)
     ) {
-      const encryptedData = encrypt(this.props.exportStorage(), this.props.password).toString();
+      const encryptedData = encrypt(
+        JSON.stringify(this.props.exportState),
+        this.props.password
+      ).toString();
       this.props.setEncryptedCache(encryptedData);
       this.props.resetAppDb();
       this.lockScreen();
       this.setState({ shouldAutoLock: false });
     }
+
+    // After decrypt with valid password, reset db and
+    if (this.state.decryptSuccess && this.state.toImport) {
+      this.props.importState(this.state.toImport);
+      this.resetEncrypted();
+    }
   }
 
-  public decryptWithPassword = async (password: string): Promise<boolean> => {
-    const { destroyEncryptedCache, encryptedDbState, importStorage } = this.props;
+  public decryptWithPassword = (password: string): void => {
+    const { encryptedDbState } = this.props;
+    if (!encryptedDbState || !encryptedDbState.data) return;
     try {
-      if (!encryptedDbState) {
-        return false;
-      }
       const passwordHash = hashPassword(password);
       // Decrypt the data and store it to the MyCryptoCache
       const decryptedData = decrypt(encryptedDbState.data as string, passwordHash);
-      const importResult = importStorage(decryptedData);
-      if (!importResult) {
-        return false;
+      if (decryptedData) {
+        this.setState({ decryptSuccess: true, toImport: decryptedData });
       }
-
-      destroyEncryptedCache();
-
-      // Navigate to the dashboard and reset inactivity timer
-      this.setState({ locked: false }, () => {
-        this.props.history.replace(ROUTE_PATHS.DASHBOARD.path);
-      });
-      this.resetInactivityTimer();
-      return true;
     } catch (error) {
       console.error(error);
-      return false;
+      this.setState({ decryptError: error });
     }
   };
 
   // Wipes encrypted data and unlocks
   public resetEncrypted = async () => {
-    const { destroyEncryptedCache } = this.props;
+    const { history, destroyEncryptedCache } = this.props;
     destroyEncryptedCache();
-    this.setState({ locked: false });
+    this.setState({ locked: false, decryptSuccess: false }, () => {
+      this.resetInactivityTimer();
+      history.replace(ROUTE_PATHS.DASHBOARD.path);
```

Comment:
why not using `history.push` ?

Reply:
Initially I was surprised, but in fact from the moment the screen is unlocked you should not be able to return to a locked state with the router so `.replace` seems adequate.

Reply:
That's a good catch ! 👍